src/ledger/fuzz.zig

```diff
@@ -0,0 +1,466 @@
+const std = @import("std");
+const sig = @import("../sig.zig");
+const build_options = @import("build-options");
+const ledger = @import("lib.zig");
+
+const ColumnFamily = sig.ledger.database.ColumnFamily;
+
+const allocator = std.heap.c_allocator;
+
+const Data = struct {
+    value: []const u8,
+};
+
+const cf1 = ColumnFamily{
+    .name = "data",
+    .Key = u64,
+    .Value = Data,
+};
+
+var dataMap = std.AutoHashMap(u32, Data).init(allocator);
+var dataKeys = std.ArrayList(u32).init(allocator);
+
+pub const BlockstoreDB = switch (build_options.blockstore_db) {
+    .rocksdb => ledger.database.RocksDB(&.{cf1}),
+    .hashmap => ledger.database.SharedHashMapDB(&.{cf1}),
+};
+
+fn createBlockstoreDB() !BlockstoreDB {
+    const ledger_path =
+        try std.fmt.allocPrint(allocator, "{s}/ledger", .{sig.FUZZ_DATA_DIR});
+
+    // ensure we start with a clean slate.
+    if (std.fs.cwd().access(ledger_path, .{})) |_| {
+        try std.fs.cwd().deleteTree(ledger_path);
+    } else |_| {}
+    try std.fs.cwd().makePath(ledger_path);
+
+    return try BlockstoreDB.open(
+        allocator,
+        .noop,
+        ledger_path,
+    );
+}
+
+pub fn run(seed: u64, args: *std.process.ArgIterator) !void {
+    const maybe_max_actions_string = args.next();
+
+    const maybe_max_actions = if (maybe_max_actions_string) |max_actions_str|
+        try std.fmt.parseInt(usize, max_actions_str, 10)
+    else
+        null;
+
+    var prng = std.rand.DefaultPrng.init(seed);
+    const random = prng.random();
+
+    const ledger_path =
+        try std.fmt.allocPrint(allocator, "{s}/ledger", .{sig.FUZZ_DATA_DIR});
+
+    // ensure we start with a clean slate.
+    if (std.fs.cwd().access(ledger_path, .{})) |_| {
+        try std.fs.cwd().deleteTree(ledger_path);
+    } else |_| {}
+    try std.fs.cwd().makePath(ledger_path);
+
+    var db = try createBlockstoreDB();
+
+    defer db.deinit();
+
+    const functions = .{
+        dbPut,
+        dbGet,
+        dbGetBytes,
+        dbCount,
+        dbContains,
+        dbDelete,
+        dbDeleteFilesInRange,
+        // Batch API
+        // - batch.put
+        // - batch.delete
+        // - batch.deleteRange
+        batchAPI,
+    };
+
+    inline for (functions) |function| {
+        const fn_args = .{ &db, random, maybe_max_actions };
+        _ = try @call(.auto, function, fn_args);
+    }
```

Comment:
i feel like it would be better to have a similar approach to accountsdb where you have a enum over all functions and then randomly select which function to run and have a single max-actions to run would be better/more random

Reply:
> i feel like it would be better to have a similar approach to accountsdb where you have a enum over all functions and then randomly select which function to run 

I considered this, but the reasoning why I did not go with that approach was because I felt the randomness should be in how the methods are called (ie the inputs passed to them) and not if the methods we are called - which would be the case if the fuzz tests randomly selects the methods to execute while fuzzing.

The advantage of going with the enum approach is for dependencies between the method calls and if randomizing that would expose any hidden bugs.

I think it should be possible to adapt what is there now to that. I'll give it a shot

Reply:
>  I felt the randomness should be in how the methods are called (ie the inputs passed to them) and not if the methods we are called

why not both? more variability here seems like a bonus

Reply:
[c695ce7](https://github.com/Syndica/sig/pull/452/commits/c695ce776cf0e94aa2a502233de86c3696196a62)